### ISSUE TITLE
Allow batch modifications to ko.observableArray via an "edit script".

### ISSUE THREAD

**Issue by ThomasMichon** [knockout/knockout#1822]
Knockout's ```arrayChange``` event is extremely useful for minimizing the number of updates made to bound data-sources. However, this only works effectively if you make individual edits to observable arrays using the wrapper ```push```, ```pop```, ```splice```, and other methods. There is no general way to replay an "edit script" into an observable array.

Suppose I want to do something simple, like keep one observable array in sync with another.

``` javascript
// You have a source array somewhere in a data model.
var array1 = ko.observableArray('a', 'b');

// Now you have a second array in a view model, to which you bind to change events.
var array2 = ko.observableArray();

// For a time, I want to keep array2 in sync with array1.
ko.computed(function () {
    array2(array1());
});

// Now the data model receives a new update:
array1('a', 'd', 'b', 'c');

// At this moment, array2 will have the correct values.
// However, array2 had to re-run compareArrays even though we already know the changes!
```

Now, I can theoretically fix performance by doing the following instead of using a ```ko.computed```:

``` javascript
array1.subscribe(function (changes) {
    var array = array2.peek();
    
    for (var i = 0; i < changes.length; i++) {
        switch (changes[i].status) {
            case 'added': array.splice(changes[i].index, 0, changes[i].value); break;
            case 'deleted': array.splice(changes[i].index, 1); break;
        }
    }

    array2.notifySubscribers(changes, 'arrayChange');
});
```

The problem is, anything just subscribing to normal change events on ```array2``` won't get them, but if I call ```array2.valueHasMutated()```, the ```trackArrayChanges``` logic will run a full diff!

Basically, if we want to support efficient piping of array changes, we need a method on ```ko.observableArray``` which takes a change list and makes the appropriate underlying edits so it can output a correct list of minimal changes. Perhaps this:

``` javascript
array1.subscribe(function (changes) {
    array2.applyChanges(changes);
});
```

This issue is similar to https://github.com/knockout/knockout/issues/1202, but in this case I'm asking to add a new method to ```ko.observableArray``` so that the whole "diff cache" concept can remain internal to the system, rather than exposed for extension.

**Comment by brianmhunt** [comment]
Thanks for the report @ThomasMichon 

I agree that this would be useful.

You may be interested in looking at what I did in [`knockout-fast-foreach`](https://github.com/brianmhunt/knockout-fast-foreach/blob/master/index.js#L109-L208)

**Comment by mbest** [comment]
This is interesting. Thanks for the suggestion.
